Useful Starting Page backend changes

adam-james-v opened the pull request:
Supersedes: #21366 because I thought some test issues required a rebase. They didn't, but this is cleaner now.

Epic: #21322 

#### BE changes

- [x] Add `is_installer` field to `/api/user/current`. It's basically the user with `id = 1`. Make sure to update `add-has-invited-second-user` check.
- [x] Add `has_question_and_dashboard` to `/api/user/current`. Should be `true` if the current user has access to at least one question *and* dashboard.
- [x] Add `/api/activity/popular_views` using the logic outlined in [the product doc](https://www.notion.so/metabase/Give-users-a-chill-and-useful-starting-page-41fce022e6c448bd8036d4facec7eb51#1e39bc3549a547b4802654fda2fb1e09). Could return the same data as from `/api/activity/recent_views`.
- [x] (Optional) Add `first_login` to `/api/user/current`. If not present in the db, this field can be substituted with `date_joined` for existing users. The field mirrors `last_login` field.

adam-james-v committed:
Add `:is_installer` key to users returned from api/user/current

adam-james-v committed:
Moved `permissions-set` function up so I can use it in hydration

adam-james-v committed:
Adjust recents query and first pass at popular items query

Recents:

Before, the recent query would look at the ViewLog, which records a view on cards even when it was only 'viewed' via a
dashboard.

For dashboards and tables, we don't need to make a change.

For cards, we now query the QueryExecution table and use started_at as a proxy for the last viewed timestamp, and can
then only grab cards with an execution context of 'question'.

Popular:

We now want to have a notion of 'popular' items as well, which can look different than 'recents'. This is the first
attempt at a reasonable query with scoring system.

The score takes into account:
 - recency of view
 - count of views total
 - if the card/dashboard is 'verified' (enterprise)
 - if the cad/dashboard is inside an 'official' Collection (enterprise)

The popular score currently uses only the *current-user-id* to fetch items, so popularity is (in this commit, at
least) a per-user concept.

adam-james-v committed:
Fixed mistake and renamed endpoint

adam-james-v committed:
Typo fix

adam-james-v committed:
Clean up QueryExecution in tests

adam-james-v committed:
Add `:has_question_and_dashboard` Key to current users via hydration

This is also part of the starting page project, it's a key that helps the frontend serve more specific data to the
user.

adam-james-v committed:
Indent, naming, and some simple defs

adam-james-v committed:
try to reduce db calls for :has_question_and_dashboard

I've moved the fn out of the hydration system to guarantee that it only runs for a single user, and is only used in
`GET /api/user/current` (and not accidentally used in other spots via hydration mechanism).

Still don't know how to check Card and Dashboard tables more efficiently

adam-james-v committed:
ViewLog and QueryExecution have different timestamps. Can't compare

Pushing this for some review, but might want to build a proper way to compare, so that recent cards and recent
dashboards/tables are sorted equally

adam-james-v committed:
fix namespace sorting

ranquild committed:
Fix the endpoint name

adam-james-v committed:
Sorting mixed date time formats 'works' but not ideal

This formats the timestamps into strings and uses those for comparison. Not perfect.

Pushing in case people are trying the branch

dpsutton commented on src/metabase/api/activity.clj:
```diff
@@ -107,31 +126,129 @@
                           (not (or (:archived model-object)
                                    (= (:visibility_type model-object) :hidden))))]
-           (cond-> (assoc view-log :model_object (dissoc model-object :dataset_query))
+           (cond-> (assoc view-log :model_object model-object)
```
why remove the `dissoc :dataset_query`?

dpsutton commented on src/metabase/api/user.clj:
```diff
@@ -161,11 +161,24 @@
+        perms-query {:where [:and
+                             [:= :archived false]
+                             coll-ids-filter]}]
+    (assoc user :has_question_and_dashboard (and (> (db/count 'Card (perms-query user)) 0)
```
you can use `toucan.db/exists?` instead of a count > 0 here.

dpsutton replied:
also, this notion seems quite strange to me. If someone had 500 cards and 0 dashboards or someone had no permissions to see anything this would report the same. Seems weird to lump cards and dashboards together like this.

adam-james-v committed:
Use simpler db function

adam-james-v committed:
Let views and runs query work with one user or all users

Popular_items are an all-users notion, but recent_views applies only to the current user.

dpsutton committed:
Unify view_log.timestamp to query_execution.started_at type

these used to both be `DATETIME` in the migration file. Then migration
168 bumped up the type of the view_log.timestamp column:

```

  - changeSet:
      id: 168
      author: camsaul
      comment: Added 0.36.0
      changes:
        - modifyDataType:
            tableName: query_execution
            columnName: started_at
            newDataType: ${timestamp_type}
```

So these were no longer the same on h2 (and possibly mysql). But sorting
viewlogs and query_executions would fail.

```clojure
activity=> (->> (ViewLog) first :timestamp #_type)
        "0x6a33e42e"
        "2022-04-04T21:57:07.471"]
activity=> (->> (ViewLog) first :timestamp type)
java.time.LocalDateTime
activity=> (->> (QueryExecution) first :started_at #_type)
        "0x7af249ac"
        "2022-04-04T21:57:07.625738Z"]
activity=> (->> (QueryExecution) first :started_at type)
java.time.OffsetDateTime
```

The LocalDateTime and OffsetDateTime were not comparable. So make them
identical types again.

adam-james-v committed:
Bookmarked results should not show up in recents/ popular.

This is done in a db query to still try fetch enough items from the db to present to the user.

Filtering bookmarked items later may result in an empty list. It's possible that the firt N results are all
bookmarked, and then the frontend would have no items to show. Filtering bookmarked results out from the beginning
increases the chances of a non-empty result.

adam-james-v committed:
:first_login populated with the earliest login timestamp.

If there is a first login timestamp, that is used. If one does not exist, we assume it's the first time logging in and
use an OffsetDateTime (now). This is possible since the login_history is done off thread, so on a real first login, we
might hit the db before the login is logged, resulting in an empty list returned.

On subsequent checks, we should see a proper timestamp from the db.

Since this is used to check if a user is 'new' (within 7 days of first logging in), the accuracy of the timestamp
matters on the order of days, not milliseconds, so this should be ok.

adam-james-v committed:
Passing test

adam-james-v committed:
Popular_items test

Tweak the create-views! function to more consistently order items. And creates views for dashboards/tables (ViewLog)
and cards (QueryExecution) in a unified way, meaning we can reliably order views, and write tests more easily.

Note that the popular_items test may have to change if the scoring math changes.

ranquild commented on src/metabase/api/activity.clj:
```diff
@@ -107,31 +126,137 @@
           (group-by :model views))))
 
+(defn- views-and-runs
+  "Common query implementation for `recent_views` and `popular_views`. Tables and Dashboards have a query limit of `views-limit`.
```
nit: `popular_items`

ranquild commented on src/metabase/api/activity.clj:
```diff
@@ -107,31 +126,137 @@
+          (assoc item :score (double (reduce + scores))))))))
+
+(defendpoint GET "/popular_items"
+  "WIP Get the list of 5 popular things for the current user. Query takes 8 and limits to 5 so that if it
```
nit: Is `WIP` intended?

ranquild reviewed (changes requested):
`has_question_and_dashboard` works incorrectly:
- it returns numbers, e.g. `0`
- it returns `0` when there is actually a question and a dashboard available to this user

ranquild committed:
Fix e2e test

adam-james-v committed:
Fix nit and bug

- forgot to remove '0' from the and clause, so we didn't get the expected boolean
- popular_items not WIP

adam-james-v committed:
another nit

ranquild committed:
Change the check in a test

adam-james-v committed:
If no views, send cards, dashboards, and tables that user *can* view

This is for a situation where there are no (or not enough) items that have views or runs. For instance, if the
instance is new and the API was used to add cards or dashboards.

The user should still see something here, so we pull from the permissions set any cards and dashboards as well as any tables.

adam-james-v committed:
Handle 'dataset' cards as well

adam-james-v committed:
Remove hard-coded ids causing tests to fail

adam-james-v committed:
DB detail editing permission enforcement (#21531)

* first pass

* a few more endpoints

* add can_access_db_details to /api/user/common

adam-james-v committed:
Simpler test to allow us to merge this, refine later

This lets us check that we get non-empty list of items, but we don't necessarily know the contents of the list or the order

ranquild committed:
Add ttag
